crates/sncast/src/response/print.rs

```diff
@@ -0,0 +1,242 @@
+        match &self {
+            OutputValue::String(s) => s.fmt(fmt),
+            OutputValue::Array(arr) => {
+                let arr_as_string = arr
+                    .iter()
+                    .map(ToString::to_string)
+                    .collect::<Vec<String>>()
+                    .join(", ");
+                write!(fmt, "[{arr_as_string}]")
+            }
+        }
+    }
+}
+
+pub fn print_command_result<T: CommandResponse>(
+    command: &str,
+    result: &mut Result<T>,
+    numbers_format: NumbersFormat,
+    output_format: &OutputFormat,
+) -> Result<()> {
+    let mut output: OutputData = vec![];
+    output.push((
+        String::from("command"),
+        OutputValue::String(command.to_string()),
+    ));
+    output.extend(result_as_output_data(result));
+    let formatted_output = output
+        .into_iter()
+        .map(|(k, v)| (k, apply_numbers_formatting(v, numbers_format)))
+        .collect();
+
+    for val in pretty_output(formatted_output, output_format)? {
+        match result {
+            Ok(_) => println!("{val}"),
+            Err(_) => eprintln!("{val}"),
+        }
+    }
+    Ok(())
+}
+
+fn pretty_output(output: OutputData, output_format: &OutputFormat) -> Result<Vec<String>> {
+    match output_format {
+        OutputFormat::Json => {
+            let json_output: HashMap<String, OutputValue> = output.into_iter().collect();
+            let pretty_string = serde_json::to_string(&json_output)?;
+            Ok(vec![pretty_string])
+        }
+        OutputFormat::Human => {
+            let mut result = vec![];
+            for (key, value) in &output {
+                let value = value.to_string();
+                result.push(format!("{key}: {value}"));
+            }
+            Ok(result)
+        }
+    }
+}
+
+fn result_as_output_data<T: CommandResponse>(result: &mut Result<T>) -> OutputData {
+    match result {
+        Ok(response) => {
+            let struct_value =
+                serde_json::to_value(response).expect("Failed to serialize CommandResponse");
+            struct_value_to_output_data(struct_value)
+        }
+        Err(message) => {
+            vec![(
+                String::from("error"),
+                OutputValue::String(format!("{message:#}")),
+            )]
+        }
+    }
+}
+
+fn struct_value_to_output_data(struct_value: Value) -> OutputData {
+    match struct_value {
+        Value::Object(obj) => obj
+            .into_iter()
+            .filter(|(_, v)| !(matches!(v, Value::Null)))
+            .map(|(k, v)| (k, value_to_output_value(v)))
+            .collect(),
+        _ => panic!("Expected an object"),
+    }
+}
+
+fn value_to_output_value(value: Value) -> OutputValue {
+    match value {
+        Value::Array(a) => OutputValue::Array(a.into_iter().map(value_to_output_value).collect()),
+        Value::String(s) => OutputValue::String(s.to_string()),
+        s => panic!("{s:?} cannot be auto-serialized to output"),
```

Comment:
philosophical question - we avoid panicing in cast an any cost basically, I know generally it is ok to do that in binaries, but I'm not sure we should break prog execution that way. Is there any special rationale to do that instead of returning Err in a manner we did up until this point? This is something I wanted to discuss on interfaces meeting, but I'm happy to briefly discuss here as well.

Reply:
Rule of thumb is:
- return `Result` when error can be caused by user input/behaviour
- panic when error should never happen and if it happend then you have wrong assumptions while writing the function so you want to know you did something incorrectly. 

Here the generated values are purely based on `CommandResonse` structures which are defined in our codebase in a way that should never parse to something including for example numbers.

Reply:
For example you should be ok using panic in cases like `"test".parse().unwrap()` because it is hardcoded. If it breaks you probably should panic because that will clearly signalise to the user that soft is broken and that is not their fault.

Reply:
sure, makes sense. we panic in tests, just avoided doing so in prod code :+1: thanks for an explanation